packages/fxa-shared/nestjs/gql/gql-whitelist.ts

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+import { readFileSync } from 'fs';
+import { Request, Response, NextFunction } from 'express';
+
+/**
+ * Configuration options for GqlGuard
+ */
+export type Config = {
+  /** A list of json files holding queries extracted with the persistgraphql util. */
+  whitelist: string[];
+  /** Toggles gaurd on / off. */
+  enabled: boolean;
+};
+
+/**
+ * Guards against the execution of unsanctioned gql queries. A list of allowed queries can be populated by running.
+ * `yarn gql:whitelist`.
+ *
+ * During development it is probably desirable to disable this!
+ */
+export class GqlWhitelist {
+  public readonly valid: Array<Record<string, number>>;
+  public readonly enabled: boolean;
+
+  constructor(config: Config) {
+    this.valid = config.whitelist.map((x) => {
+      const raw = readFileSync(x).toString();
+      try {
+        return JSON.parse(raw);
+      } catch (err) {
+        console.warn('Failed to parse gql whitelist', raw);
+      }
+    });
+    this.enabled = config.enabled;
+  }
+
+  allowed(req: Pick<Request, 'body' | 'query'>) {
+    const query = req.query?.query || req.body?.query;
+
+    // If there was no query defined, assume it is not a gql request.
+    // This allows the gql playground to work.
+    if (!query) {
+      return true;
+    }
+
+    return this.valid.some((x) => x[query] != null);
```

Comment:
well this is gonna be nitpicky... I feel I haven't seen `!=` in a while (we don't lint for non-strict comparisons eh?), so maybe `x[query] !== undefined` in the strict direction or just `!!x[query]` in the falsy direction?

Reply:
BTW, this is idiomatic js to make sure a value is not null or undefined. I feel like I see it used quite a bit. In this case the the null scenario is probably not a thing we actually need to check, but... I've certainly been bitten by that assumption more than once.  Note that `!!` would also evaluate false for an empty string, but I think that's probably fine in this case too.

Reply:
> ... make sure a value is not null or undefined ...

Yeah that make sense.  If you want to use a single operator that's literally the only that would work.

Reply:
Looks like we currently have 69 combined usages of `!= null` (27) and `== null` (42):

```sh
git grep -IPn " (!=|==) null" -- ':!packages/123done' ':!.yarn/' ':!*.sh' ':!*.yml' ':!*.py' | wc -l # 69
git grep -IPn " (!=|==) null" -- ':!packages/123done' ':!.yarn/' ':!*.sh' ':!*.yml' ':!*.py' | grep " == null" | wc -l # 42
git grep -IPn " (!=|==) null" -- ':!packages/123done' ':!.yarn/' ':!*.sh' ':!*.yml' ':!*.py' | grep " != null" | wc -l # 27
```

Reply:
> (we don't lint for non-strict comparisons eh?)

Ah, I missed this comment earlier. The ESLint rule in question is [`eqeqeq`](https://eslint.org/docs/latest/rules/eqeqeq):

```sh
git grep -n "eqeqeq" | cat

packages/eslint-plugin-fxa/lib/configs/default.js:24:    eqeqeq: [2, 'allow-null'],
```

But per the docs, it looks like our config is deprecated:
> ### [`allow-null`](https://eslint.org/docs/latest/rules/eqeqeq#allow-null)
> **Deprecated:** Instead of using this option use `"always"` and pass a `"null"` option property with value `"ignore"`. This will tell ESLint to always enforce strict equality except when comparing with the `null` literal.
>
> ```js
> ["error", "always", {"null": "ignore"}]
> ```